File: code/0015. 3Sum/solution4.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        map<int, int> numCount;
        vector<vector<int>> result;
        
        for (int num : nums) {
            numCount[num]++;
        }
        
        
        for (auto it = numCount.begin(); it != numCount.end(); it++) {
            int num1 = (*it).first;
            (*it).second = (*it).second - 1;
            for (auto it2 = it; it2 != numCount.end(); it2++) {
                if ((*it2).second > 0) {
                    int num2 = (*it2).first;
                    if (num2 > -num1 - num2) {
                        break;
                    }
                    (*it2).second = (*it2).second - 1;
                    if (numCount.count(-num1 - num2) > 0 && numCount[-num1 - num2] > 0) {
                        result.push_back({num1, num2, -num1 - num2});
                    }
                    (*it2).second = (*it2).second + 1;
                }                
            }
            (*it).second = (*it).second + 1;
        }
        
        return result;
    }
};

File: code/0015. 3Sum/solution4.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Map<Integer, Integer> numCount = new TreeMap();
        for (int num : nums) {
            numCount.put(num, numCount.getOrDefault(num, 0) + 1);
        }
        
        List<Integer> sortedNums = new ArrayList<>(numCount.keySet());
        
        for (int i = 0; i < sortedNums.size(); i++) {
            int num1 = sortedNums.get(i);
            numCount.put(num1, numCount.get(num1) - 1);
            for (int j = i; j < sortedNums.size(); j++) {
                int num2 = sortedNums.get(j);
                if (num2 > -num1 - num2) {
                    break;
                }
                if (numCount.get(num2) > 0) {
                    numCount.put(num2, numCount.get(num2) - 1);
                    if (numCount.getOrDefault(-num1 - num2, 0) > 0) {
                        result.add(Arrays.asList(num1, num2, -num1 - num2));
                    }
                    numCount.put(num2, numCount.get(num2) + 1);
                }
            }
            numCount.put(num1, numCount.get(num1) + 1);
        }
        
        return result;
    }
}